src/ralph/backends/database/mongo.py

```diff
@@ -0,0 +1,128 @@
+"""MongoDB database backend for Ralph"""
+
+import hashlib
+import json
+import logging
+from typing import Generator, TextIO, Union
+
+from bson.objectid import ObjectId
+from pymongo import MongoClient
+from pymongo.errors import BulkWriteError
+
+from ralph.defaults import MONGO_COLLECTION, MONGO_DATABASE, MONGO_URI
+
+from .base import BaseDatabase
+
+logger = logging.getLogger(__name__)
+
+
+class MongoDatabase(BaseDatabase):
+    """Mongo database backend."""
+
+    name = "mongo"
+
+    def __init__(
+        self,
+        connection_uri: str = MONGO_URI,
+        database: str = MONGO_DATABASE,
+        collection: str = MONGO_COLLECTION,
+        client_options: dict = None,
+    ):
+        """Instanciates the Mongo client.
+
+        Args:
+            connection_uri (str): MongoDB connection URI.
+            database (str): MongoDB database to connect to.
+            collection (str): MongoDB database collection to get objects from.
+            client_options (dict): A dictionary of valid options for the MongoClient
+                class initialization.
+        """
+        if client_options is None:
+            client_options = {}
+
+        self.client = MongoClient(connection_uri, **client_options)
+        self.database = getattr(self.client, database)
+        self.collection = getattr(self.database, collection)
+
+    def get(self, chunk_size: int = 500):
+        """Gets collection documents and yields them."""
+
+        for document in self.collection.find(batch_size=chunk_size):
+            # Make the document json-serializable
+            document.update({"_id": str(document.get("_id"))})
+            yield document
+
+    @staticmethod
+    def to_documents(stream: Union[TextIO, list]) -> Generator[dict, None, None]:
+        """Converts `stream` lines (one statement per line) to Mongo documents.
+
+        We expect statements to have at least an `id` field that will be used
+        to compute a unique MongoDB Object ID. This ensures that we will not
+        duplicate statements in our database.
+        """
+
+        for line in stream:
+            statement = json.loads(line) if isinstance(line, str) else line
+            document = {
+                "_id": ObjectId(
+                    hashlib.sha256(
+                        bytes(statement.get("id", None), "utf-8")
+                    ).hexdigest()[:24]
+                ),
+                "statement": statement,
```

Comment:
Why do we put the `statement` inside a "statement" key? Could we write the input data directly at the root level?

Reply:
I mimicked learning locker's schema: https://docs.learninglocker.net/http-statements/

Reply:
Could the MongoDB backend be more generic in this regard? For example, [lxHive LRS](https://github.com/g3i/lxHive) seems to store xAPI statements at the root level of the MongoDB document.

Reply:
That's interesting! We should had this to our future IRL discussion.